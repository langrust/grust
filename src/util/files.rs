use codespan_reporting::files;
use std::ops::Range;

/// A file that is backed by an `Arc<String>`.
#[derive(Debug, Clone)]
struct File {
    /// The name of the file.
    name: String,
    /// The source code of the file.
    source: String,
    /// The starting byte indices in the source code.
    line_starts: Vec<usize>,
}

impl File {
    fn line_start(&self, line_index: usize) -> Result<usize, files::Error> {
        use std::cmp::Ordering;

        match line_index.cmp(&self.line_starts.len()) {
            Ordering::Less => Ok(*self
                .line_starts
                .get(line_index)
                .expect("failed despite previous check")),
            Ordering::Equal => Ok(self.source.len()),
            Ordering::Greater => Err(files::Error::LineTooLarge {
                given: line_index,
                max: self.line_starts.len() - 1,
            }),
        }
    }
}

/// An opaque file identifier.
#[derive(Copy, Clone, PartialEq, Eq, Debug)]
pub struct FileId(i32);
impl FileId {
    /// Construct and return the default file identifier.
    pub fn default() -> Self {
        return FileId(-1);
    }
}

/// The files handler.
#[derive(Debug, Clone)]
pub struct Files {
    files: Vec<File>,
}

impl Files {
    /// Create a new files database.
    pub fn new() -> Files {
        Files { files: Vec::new() }
    }

    /// Add a file to the database, returning the handle that can be used to
    /// refer to it again.
    pub fn add(
        &mut self,
        name: impl Into<String>,
        source: impl Into<String>,
    ) -> Option<FileId> {
        let file_id = FileId(i32::try_from(self.files.len()).ok()?);
        let name = name.into();
        let source = source.into();
        let line_starts = files::line_starts(&source).collect();

        self.files.push(File {
            name,
            line_starts,
            source,
        });

        Some(file_id)
    }

    /// Get the file corresponding to the given id.
    fn get(&self, file_id: FileId) -> Result<&File, files::Error> {
        self.files
            .get(file_id.0 as usize)
            .ok_or(files::Error::FileMissing)
    }
}

impl<'files> files::Files<'files> for Files {
    type FileId = FileId;
    type Name = &'files str;
    type Source = &'files str;

    fn name(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(self.get(file_id)?.name.as_ref())
    }

    fn source(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(&self.get(file_id)?.source)
    }

    fn line_index(
        &self,
        file_id: FileId,
        byte_index: usize
    ) -> Result<usize, files::Error> {
        self.get(file_id)?
            .line_starts
            .binary_search(&byte_index)
            .or_else(|next_line| Ok(next_line - 1))
    }

    fn line_range(
        &self,
        file_id: FileId,
        line_index: usize,
    ) -> Result<Range<usize>, files::Error> {
        let file = self.get(file_id)?;
        let line_start = file.line_start(line_index)?;
        let next_line_start = file.line_start(line_index + 1)?;

        Ok(line_start..next_line_start)
    }
}