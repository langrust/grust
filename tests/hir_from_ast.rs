use codespan_reporting::files::SimpleFiles;

use grustine::{error::display, hir_from_ast};

#[test]
fn hir_from_ast_transformation_for_counter() {
    let mut files = SimpleFiles::new();

    let counter_id = files.add(
        "counter.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/counter.gr").expect("unkown file"),
    );

    match hir_from_ast(counter_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_blinking() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "blinking.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/blinking.gr").expect("unkown file"),
    );

    match hir_from_ast(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_button_management() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "button_management.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/button_management.gr").expect("unkown file"),
    );

    match hir_from_ast(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_button_management_condition_match() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "button_management_condition_match.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/button_management_condition_match.gr")
            .expect("unkown file"),
    );

    match hir_from_ast(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_button_management_using_function() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "button_management_using_function.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/button_management_using_function.gr")
            .expect("unkown file"),
    );

    match hir_from_ast(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_pid() {
    let mut files = SimpleFiles::new();

    let pid_id = files.add(
        "pid.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/pid.gr").expect("unkown file"),
    );

    match hir_from_ast(pid_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}
