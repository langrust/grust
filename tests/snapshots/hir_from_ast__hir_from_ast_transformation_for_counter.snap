---
source: tests/hir_from_ast.rs
expression: file
---
typedefs: []
functions: []
nodes:
  - id: counter
    is_component: false
    inputs:
      - - res
        - Boolean
      - - tick
        - Boolean
    unscheduled_equations:
      inc:
        scope: Local
        id: inc
        signal_type: Integer
        expression:
          FunctionApplication:
            function_expression:
              Call:
                id: if_then_else
                typing:
                  Abstract:
                    - - Boolean
                      - Integer
                      - Integer
                    - Integer
                location: ~
            inputs:
              - SignalCall:
                  signal:
                    id: tick
                    scope: Input
                  typing: Boolean
                  location: ~
                  dependencies: ~
              - Constant:
                  constant:
                    Integer: 1
                  typing: Integer
                  location: ~
                  dependencies: ~
              - Constant:
                  constant:
                    Integer: 0
                  typing: Integer
                  location: ~
                  dependencies: ~
            typing: Integer
            location: ~
            dependencies: ~
        location: ~
      o:
        scope: Output
        id: o
        signal_type: Integer
        expression:
          FunctionApplication:
            function_expression:
              Call:
                id: if_then_else
                typing:
                  Abstract:
                    - - Boolean
                      - Integer
                      - Integer
                    - Integer
                location: ~
            inputs:
              - SignalCall:
                  signal:
                    id: res
                    scope: Input
                  typing: Boolean
                  location: ~
                  dependencies: ~
              - Constant:
                  constant:
                    Integer: 0
                  typing: Integer
                  location: ~
                  dependencies: ~
              - FunctionApplication:
                  function_expression:
                    Call:
                      id: " + "
                      typing:
                        Abstract:
                          - - Integer
                            - Integer
                          - Integer
                      location: ~
                  inputs:
                    - FunctionApplication:
                        function_expression:
                          Call:
                            id: (_)
                            typing:
                              Abstract:
                                - - Integer
                                - Integer
                            location: ~
                        inputs:
                          - FollowedBy:
                              constant:
                                Integer: 0
                              expression:
                                SignalCall:
                                  signal:
                                    id: o
                                    scope: Output
                                  typing: Integer
                                  location: ~
                                  dependencies: ~
                              typing: Integer
                              location: ~
                              dependencies: ~
                        typing: Integer
                        location: ~
                        dependencies: ~
                    - SignalCall:
                        signal:
                          id: inc
                          scope: Local
                        typing: Integer
                        location: ~
                        dependencies: ~
                  typing: Integer
                  location: ~
                  dependencies: ~
            typing: Integer
            location: ~
            dependencies: ~
        location: ~
    unitary_nodes: {}
    contracts:
      requires: []
      ensures: []
      invariant: []
      assert: []
    location: ~
    graph: ~
component:
  id: main
  is_component: true
  inputs: []
  unscheduled_equations:
    half:
      scope: Local
      id: half
      signal_type: Boolean
      expression:
        FollowedBy:
          constant:
            Boolean: true
          expression:
            FunctionApplication:
              function_expression:
                Call:
                  id: "!"
                  typing:
                    Abstract:
                      - - Boolean
                      - Boolean
                  location: ~
              inputs:
                - SignalCall:
                    signal:
                      id: half
                      scope: Local
                    typing: Boolean
                    location: ~
                    dependencies: ~
              typing: Boolean
              location: ~
              dependencies: ~
          typing: Boolean
          location: ~
          dependencies: ~
      location: ~
    y:
      scope: Output
      id: y
      signal_type: Integer
      expression:
        NodeApplication:
          node: counter
          inputs:
            - FollowedBy:
                constant:
                  Boolean: false
                expression:
                  FunctionApplication:
                    function_expression:
                      Call:
                        id: (_)
                        typing:
                          Abstract:
                            - - Boolean
                            - Boolean
                        location: ~
                    inputs:
                      - FunctionApplication:
                          function_expression:
                            Call:
                              id: " > "
                              typing:
                                Abstract:
                                  - - Integer
                                    - Integer
                                  - Boolean
                              location: ~
                          inputs:
                            - SignalCall:
                                signal:
                                  id: y
                                  scope: Output
                                typing: Integer
                                location: ~
                                dependencies: ~
                            - Constant:
                                constant:
                                  Integer: 35
                                typing: Integer
                                location: ~
                                dependencies: ~
                          typing: Boolean
                          location: ~
                          dependencies: ~
                    typing: Boolean
                    location: ~
                    dependencies: ~
                typing: Boolean
                location: ~
                dependencies: ~
            - SignalCall:
                signal:
                  id: half
                  scope: Local
                typing: Boolean
                location: ~
                dependencies: ~
          signal: o
          typing: Integer
          location: ~
          dependencies: ~
      location: ~
  unitary_nodes: {}
  contracts:
    requires: []
    ensures: []
    invariant: []
    assert: []
  location: ~
  graph: ~
location: ~

